### ISSUE TITLE
iframe gets messed after submit

### ISSUE THREAD

**Issue by haggen** [tiff/wysihtml5#101]
I just started using wysihtml5 on my client's project. It will requires the feature to embed videos. I got everything working but when I  submit my form `<iframe>` gets removed; I changed the rules to set `remove=0` and then tried again. The elements stays but its attributes are removed. There's no other rule about iframes, I checked.

How do I get this working ? I really need this and I'm already a little late.

Do I have to whitelist iframe's attributes or something ?

**Comment by tiff** [comment]
Try:

``` javascript
        "iframe": {
            "check_attributes": {
                "src": "url"
            }
        },
```

Attributes need to be white-listed, too.

**Comment by haggen** [comment]
But what about attr like `allowtransparency` that takes no value ? And
other attr that doesnt fit in numeric, alt, or url patterns ?

**Comment by tiff** [comment]
Good point. I'm considering to make the parser rules much simpler by allowing regular expressions for attribute-checks.

**Comment by haggen** [comment]
Actually, I'm gonna do the hack and send a pull request. :)

I sent a pull-request enabling boolean attribute values. Regexp would be nice, but IMO one of the great feats of this RTE is that it's simpler and safer by hiding the complexity (aka not giving too much freedom). Regex would allow too many mistakes, no ?

**Comment by khodzha** [comment]
tiff, help me please too :)
I'm trying to embed youtube video, inserting html like

``` html
<iframe width="560" height="315" src="http://www.youtube.com/embed/qkTyYC75f4k" frameborder="0" allowfullscreen></iframe>
```

and my parsing rules are

``` javascript
        "iframe": {
            "check_attributes": {
                "src":"url"
            },
            "set_attributes": {
                "width": "560",
                "height": "315",
                "frameborder": "0",
                "allowfullscreen": ""
            }
        }
```

but src attribute is removed and src attribute value appears inside `<iframe>_here_</iframe>`. What am I doing wrong? It seems link I should white-list src attribute of iframe, but how can I do it?

**Comment by haggen** [comment]
Check if you're using latest version and use `href` instead of `url` for `src`. Also I think `allowfullscreen` won't work. I made a pull request with support for booleans, you can check out my commit: https://github.com/haggen/wysihtml5/commits/master

**Comment by khodzha** [comment]
"href" doesn't work at all, so I returned to "url". But I found interesting thing - when I paste the `<iframe>` code when switched to html view - everything is ok, but pasting when default view is enabled breakes the `<iframe>`, removing src attribute :\

**Comment by haggen** [comment]
`href` was recently added, did you check if you got it ? I really don't know what else to do.

Anyway, the thing is wysihtml5 still in version 0.3.\* and I'm having my share of issues too. Despite the project being amazing and one of the best I've found so far it still got many flaws.

When you say "default view" and "html view" you're saying about source/preview ?

**Comment by khodzha** [comment]
My wysihtml version is 0.3.0, but I haven't checked for `href`, maybe it's fail :) And you're right - source and preview. But while `iframe` doesn't break in source view - I will use source view, it's not the most uncomfortable thing in the world

**Comment by haggen** [comment]
Don't update to the latest version, it has a serious issue: #131

**Comment by haggen** [comment]
@khodzha You may update it now, the issue is fixed. Try this and see if your problem gets solved.

**Comment by m4tm4t** [comment]
i'm running the latest live and "href" doesn't work on iframe. Also, img src no longer work even with "href", "url", "src" rule (local image /path/to/image.jpg)

**Comment by JeremyCade** [comment]
Does anyone have a working solution

**Comment by zrod** [comment]
"url" works just fine, are you guys sure this isn't related to X-Frame-Options on Chrome?

**Comment by nagi** [comment]
If anyone needs a quick and dirty hack to get this working quickly, you can modify the parser. This example ensures iframes pointing to YouTube videos are left alone:

``` javascript
  function parse(elementOrHtml, rules, context, cleanUp) {
    // ~~~ snip ~~~

    while (element.firstChild) {
      firstChild  = element.firstChild;
      element.removeChild(firstChild);
      newNode = _convert(firstChild, cleanUp);

      if (newNode) {
        if(firstChild.nodeName.toUpperCase() === "IFRAME" && firstChild.hasAttribute('src') && firstChild.getAttribute('src').match('^//www.youtube.com/embed/')) {
          fragment.appendChild(firstChild);
        } else {
          fragment.appendChild(newNode);
        }
      }
    }
  // ~~~ snip ~~~
  function _convert(oldNode, cleanUp) {
    // ~~~ snip ~~~
    for (i=0; i<oldChildsLength; i++) {
      oldChild = oldChilds[i];
      console.log(oldChild)
      if(typeof(oldChild) !== "undefined" && oldChild.nodeName.toUpperCase() === "IFRAME" && oldChild.hasAttribute('src') && oldChild.getAttribute('src').match('^//www.youtube.com/embed/')) {
        newNode.appendChild(oldChild);
      } else {
        newChild = _convert(oldChilds[i], cleanUp);
        if (newChild) {
          newNode.appendChild(newChild);
        }     
      }
    }
```
